File: oldapi/application/api/controller/Update.php
<?php
namespace app\api\controller;
use think\Request;
use think\Db;

class Update
{
    public function index(){
		return json(['code'=>400,'msg'=>'接口错误','api'=>""]);
	}
	
	//热歌榜更新
	public function Music_hot(){
		$url = 'https://www.gqink.cn/usr/themes/handsome/libs/Get.php?type=collect&media=netease&id=3778678';
		$res = @file_get_contents($url);
		$list = json_decode(trim($res,chr(239).chr(187).chr(191)),true);
		$data = [];
		for ($i = 0; $i < count($list); $i++) {
			$result = @m_163($list[$i]['song_id'])['lrc'];
			$data[] = [
				'name' => $list[$i]['name'],
				'url' => 'https://music.163.com/song/media/outer/url?id='.$list[$i]['song_id'],
				'cover' => $list[$i]['cover'],
				'author' => $list[$i]['author'],
				'lrc' => $result,
			];
		}
		$json = json_encode($data);
		if(is_array($data)){
			   $filename = 'public/hot.json';
			   $fp= fopen($filename, "w");  //w是写入模式，文件不存在则创建文件写入。
			   $len = fwrite($fp, $json);
			   fclose($fp);
			   print $len .'更新成功';
		}else {
			echo "更新失败";
		}
	}
	
	
	//更新一言
	public function Word(){
		$res = @file_get_contents('https://v1.hitokoto.cn/');
		$GET_JSON = json_decode(trim($res,chr(239).chr(187).chr(191)),true);
		if(is_array(@$GET_JSON)){
			$result = Db::table('word')->where('text',@$GET_JSON['hitokoto'])->find();
			if($result == null){
				$result = Db::table('word')->data(['text'=>@$GET_JSON['hitokoto'],'love'=>520])->insert();
				if($result){
					echo "更新成功";
				}else{
					echo "更新失败";
				}
			}else{
				echo "数据重复";
			}
		}else{
			echo "内部错误";
		}
	}	
	
	public function Wordtime(){
	    if(!empty(Request::instance()->param('time'))){
	        $time = Request::instance()->param('time');
	    }else{
	        $time = str_replace('/', '-', date("Y/m/d"));
	    }
		$url = 'http://www.dutangapp.cn/u/toxic?date='.$time;
		function httpGet($url) {
		    $curl = curl_init();
		    $httpheader[] = "Accept:*/*";
		    $httpheader[] = "Accept-Language:zh-CN,zh;q=0.8";
		    $httpheader[] = "Connection:close";
		    curl_setopt($curl, CURLOPT_USERAGENT, "Mozilla/5.0 (Windows; U; Windows NT 5.1; rv:1.7.3) Gecko/20041001 Firefox/0.10.1" );
		    curl_setopt($curl, CURLOPT_HTTPHEADER, $httpheader);
		    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
		    curl_setopt($curl, CURLOPT_TIMEOUT, 3);
		    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
		    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
		    curl_setopt($curl, CURLOPT_URL, $url);
		    $res = curl_exec($curl);
		    curl_close($curl);
		    return $res;
		}
		$GET_JSON = json_decode(trim(httpGet($url),chr(239).chr(187).chr(191)),true);
		if(@count(@$GET_JSON['data'])){
		    for ($i = 0; $i < count($GET_JSON['data']); $i++) {
		        $result = Db::table('word')->where('text',@$GET_JSON['data'][$i]['data'])->find();
    			if($result == null){
    				$result = Db::table('word')->data(['text'=>@$GET_JSON['data'][$i]['data'],'love'=>520])->insert();
    				if($result){
    					echo "更新成功<br/>";
    				}else{
    					echo "更新失败<br/>";
    				}
    			}else{
    				echo "数据重复<br/>";
    			}
		    }
		}else{
			echo "内部错误";
		}
	}
	
	//更新热评
	public function netease(){
		$GET_JSON = GET_JSON('https://api.uomg.com/api/comments.163');
		if(is_array(@$GET_JSON)){
			$result = Db::table('hot')->where('comment_content',@$GET_JSON['data']['content'])->find();
			if($result == null){
				$name = @$GET_JSON['data']['name'];
				$images = @$GET_JSON['data']['picurl'];
				$author = @$GET_JSON['data']['artistsname'];
				$mp3_url = @$GET_JSON['data']['url'];
				$comment_nickname = @$GET_JSON['data']['nickname'];
				$comment_content = @$GET_JSON['data']['content'];
				$love = 520;
				$result = Db::table('hot')->data(['name'=>$name,'images'=>$images,'author'=>$author,'mp3_url'=>$mp3_url,'comment_nickname'=>$comment_nickname,'comment_content'=>$comment_content,'love'=>$love])->insert();
				if($result){
					echo "更新成功";
				}else{
					echo "更新失败";
				}
			}else{
				echo "数据重复";
			}
		}else{
			echo "内部错误";
		}
	}
	
	//更新头像
	public function headimg(){
	    $kw = Request::instance()->param('kw');
	    $kwn = Request::instance()->param('kwn');
	    $a = Request::instance()->param('start');
	    $b = $a+30;//recommend
	    $url = 'http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id='.$kwn.'&start='.$a.'&len=30';
		$GET_JSON = GET_JSON($url);
		if(is_array(@$GET_JSON)){
		    $data = 0;
		    for ($i = 0; $i < count(@$GET_JSON['data']); $i++) {
		         for ($a = 0; $a < count(@$GET_JSON['data'][$i]['image']); $a++) {
                    $result = Db::table($kw)->where('url',@$GET_JSON['data'][$i]['image'][$a]['image_url'])->find();
        			if($result == null){
        				$result = Db::table($kw)->data(['url'=>@$GET_JSON['data'][$i]['image'][$a]['image_url']])->insert();
        				if($result){
        				    $data = $data+1;
        				}else{
        				    
        				}
        			}else{
        				
        			}
		         }
		    }
		    echo "<title>写入".$data."条成功</title>";
		    echo "写入".$data."条成功";
		}else{
			echo "内部错误";
		}
		echo "<script>
		window.onload=function(){
		    window.location.replace('https://api.gqink.cn/api/update/headimg/kw/$kw/kwn/$kwn/start/$b/');
		}
	   </script>";
	}
	
	//words
	public function words(){
	    $a = Request::instance()->param('start');
	    $b = $a+100;//recommend
	    $url = 'http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id=1070092&start='.$a.'&len=100';
		$GET_JSON = GET_JSON($url);
		if(is_array(@$GET_JSON)){
		    $data = 0;
	         for ($a = 0; $a < count(@$GET_JSON['data']); $a++) {
                $result = Db::table("word")->where('text',@$GET_JSON['data'][$a]['text'])->find();
    			if($result == null){
    				$result = Db::table("word")->data(['text'=>@$GET_JSON['data'][$a]['text']])->insert();
    				if($result){
    				    $data = $data+1;
    				}else{
    				    
    				}
    			}else{
    				
    			}
		    }
		    echo "<title>写入".$data."条成功</title>";
		    echo "写入".$data."条成功";
		}else{
			echo "内部错误";
		}
		echo "<script>
		window.onload=function(){
		    window.location.replace('https://api.gqink.cn/api/update/words/start/$b/');
		}
	   </script>";
	}
	
	//words
	public function qianming(){
	    $a = Request::instance()->param('start');
	    $b = $a+100;//recommend
	    $url = 'http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id=1019&start='.$a.'&len=100';
		$GET_JSON = GET_JSON($url);
		if(is_array(@$GET_JSON)){
		    $data = 0;
	         for ($a = 0; $a < count(@$GET_JSON['data']); $a++) {
                $result = Db::table("autograph")->where('word',@$GET_JSON['data'][$a]['text'])->find();
    			if($result == null){
    				$result = Db::table("autograph")->data(['word'=>@$GET_JSON['data'][$a]['text'],'keyword'=>"伤感"])->insert();
    				if($result){
    				    $data = $data+1;
    				}else{
    				    
    				}
    			}else{
    				
    			}
		    }
		    echo "<title>写入".$data."条成功</title>";
		    echo "写入".$data."条成功";
		}else{
			echo "内部错误";
		}
		echo "<script>
		window.onload=function(){
		    window.location.replace('https://api.gqink.cn/api/update/qianming/start/$b/');
		}
	   </script>";
	}
	
	public function wangming(){
	    $a = Request::instance()->param('start');
	    $b = $a+100;//recommend
	    $url = 'http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id=1025&start='.$a.'&len=100';
		$GET_JSON = GET_JSON($url);
		if(is_array(@$GET_JSON)){
		    $data = 0;
	         for ($a = 0; $a < count(@$GET_JSON['data']); $a++) {
                $result = Db::table("name")->where('word',@$GET_JSON['data'][$a]['text'])->find();
    			if($result == null){
    				$result = Db::table("name")->data(['word'=>@$GET_JSON['data'][$a]['text'],'keyword'=>"女生"])->insert();
    				if($result){
    				    $data = $data+1;
    				}else{
    				    
    				}
    			}else{
    				
    			}
		    }
		    echo "<title>写入".$data."条成功</title>";
		    echo "写入".$data."条成功";
		}else{
			echo "内部错误";
		}
		echo "<script>
		window.onload=function(){
		    window.location.replace('https://api.gqink.cn/api/update/wangming/start/$b/');
		}
	   </script>";
	}
	
	public function net(){
		require 'extend/netease/GetWangYiYunInfo.php';
        $res = new \GetWangYiYunInfo();
        $data = $res->hot('http://music.163.com/discover/toplist?id=3778678');
        for($a = 0 ; $a < count($data); $a++){
            $result = Db::table('hot')->where('comment_content',@$data[$a]['hotComments'])->find();
    		if($result == null){
    			$name = $data[$a]['song_name'];
    			$images = $data[$a]['cover'];
    			$author = $data[$a]['author'];
    			$mp3_url = $data[$a]['mp3_url'];
    			$comment_nickname = $data[$a]['nickname'];
    			$comment_content = $data[$a]['hotComments'];
    			$love = 520;
    			$result = Db::table('hot')->data(['name'=>$name,'images'=>$images,'author'=>$author,'mp3_url'=>$mp3_url,'comment_nickname'=>$comment_nickname,'comment_content'=>$comment_content,'love'=>$love])->insert();
    			if($result){
    				echo "更新成功";
    			}else{
    				echo "更新失败";
    			}
    		}else{
    			echo "数据重复";
    		}
        }
	}
	
	public function dog(){
	    $a = Request::instance()->param('start');
	    $b = $a+100;//recommend
	    $url = 'https://v1.alapi.cn/api/dog?format=json';
		$GET_JSON = GET_JSON($url);
		if(is_array(@$GET_JSON)){
			if(@$GET_JSON['code'] == 200){
				$result = Db::table("dog")->where('text',@$GET_JSON['data']['content'])->find();
				var_dump($result);
				if($result == null){
    				$result = Db::table("dog")->data(['text'=>@$GET_JSON['data']['content']])->insert();
    				if($result){
    				    echo "写入成功";
    				}else{
    				    echo "失败";
    				}
    			}else{
    				echo "重复";
    			}
			}
		}else{
			echo "内部错误";
		}
		echo "<script>
		window.onload=function(){
		    window.location.replace('https://api.gqink.cn/api/update/dog/');
		}
	   </script>";
	}
}

File: rustapi/src/handlers/api/update.rs
use std::{collections::HashMap, path::PathBuf};

use axum::{
    extract::{Query, State},
    http::{header, StatusCode},
    response::{IntoResponse, Response},
};
use chrono::Local;
use serde_json::{json, Value};
use sqlx::{MySqlPool, Row};

use crate::{respond, services, state::AppState};

type Params = HashMap<String, String>;

pub async fn index() -> Response {
    respond::json_response(json!({"code": 400, "msg": "接口错误", "api": ""}))
}

pub async fn music_hot(State(state): State<AppState>) -> Response {
    let Some(list) = fetch_json(
        &state,
        "https://www.gqink.cn/usr/themes/handsome/libs/Get.php?type=collect&media=netease&id=3778678",
    )
    .await
    .and_then(|v| v.as_array().cloned())
    else {
        return text("更新失败");
    };

    let mut data = Vec::new();
    for item in list {
        let song_id = value_string(item.get("song_id")).unwrap_or_default();
        let lrc = services::music::m_163(&state.http, &song_id)
            .await
            .and_then(|v| v.get("lrc").and_then(Value::as_str).map(ToOwned::to_owned))
            .unwrap_or_default();
        data.push(json!({
            "name": value_string(item.get("name")).unwrap_or_default(),
            "url": format!("https://music.163.com/song/media/outer/url?id={song_id}"),
            "cover": value_string(item.get("cover")).unwrap_or_default(),
            "author": value_string(item.get("author")).unwrap_or_default(),
            "lrc": lrc,
        }));
    }

    let payload = match serde_json::to_string(&data) {
        Ok(payload) => payload,
        Err(_) => return text("更新失败"),
    };
    let path = PathBuf::from(&state.config.public_dir).join("hot.json");
    if let Some(parent) = path.parent() {
        let _ = tokio::fs::create_dir_all(parent).await;
    }
    match tokio::fs::write(path, payload.as_bytes()).await {
        Ok(_) => text(format!("{}更新成功", payload.len())),
        Err(_) => text("更新失败"),
    }
}

pub async fn word(State(state): State<AppState>) -> Response {
    let Some(hitokoto) = fetch_json(&state, "https://v1.hitokoto.cn/")
        .await
        .and_then(|v| value_string(v.get("hitokoto")))
    else {
        return text("内部错误");
    };
    text(
        insert_unique(
            &state.pool,
            "word",
            "text",
            &hitokoto,
            &[("love", json!(520))],
        )
        .await,
    )
}

pub async fn wordtime(State(state): State<AppState>, Query(params): Query<Params>) -> Response {
    let day = params
        .get("time")
        .filter(|v| !v.trim().is_empty())
        .cloned()
        .unwrap_or_else(|| Local::now().format("%Y-%m-%d").to_string());
    let url = format!("http://www.dutangapp.cn/u/toxic?date={day}");
    let Some(items) = fetch_json(&state, &url)
        .await
        .and_then(|v| v.get("data").and_then(Value::as_array).cloned())
    else {
        return text("内部错误");
    };

    let mut body = String::new();
    for item in items {
        let line = value_string(item.get("data")).unwrap_or_default();
        if line.is_empty() {
            continue;
        }
        body.push_str(
            &insert_unique(&state.pool, "word", "text", &line, &[("love", json!(520))]).await,
        );
        body.push_str("<br/>");
    }
    text(if body.is_empty() {
        "内部错误".into()
    } else {
        body
    })
}

pub async fn netease(State(state): State<AppState>) -> Response {
    let Some(data) = fetch_json(&state, "https://api.uomg.com/api/comments.163")
        .await
        .and_then(|v| v.get("data").cloned())
    else {
        return text("内部错误");
    };
    text(insert_hot_comment(&state.pool, &data).await)
}

pub async fn headimg(State(state): State<AppState>, Query(params): Query<Params>) -> Response {
    let kw = params.get("kw").cloned().unwrap_or_default();
    let kwn = params.get("kwn").cloned().unwrap_or_default();
    let start = params
        .get("start")
        .and_then(|v| v.parse::<i64>().ok())
        .unwrap_or(0);
    let next = start + 30;
    let url = format!(
        "http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id={kwn}&start={start}&len=30"
    );
    let inserted = match fetch_json(&state, &url).await {
        Some(v) => {
            let mut inserted = 0;
            if let Some(items) = v.get("data").and_then(Value::as_array) {
                for item in items {
                    if let Some(images) = item.get("image").and_then(Value::as_array) {
                        for image in images {
                            if let Some(url) = value_string(image.get("image_url")) {
                                if insert_unique(&state.pool, &kw, "url", &url, &[]).await
                                    == "更新成功"
                                {
                                    inserted += 1;
                                }
                            }
                        }
                    }
                }
                format!("写入{inserted}条成功")
            } else {
                "内部错误".to_string()
            }
        }
        None => "内部错误".to_string(),
    };
    text(format!(
        "<title>{inserted}</title>{inserted}<script>window.onload=function(){{window.location.replace('https://api.gqink.cn/api/update/headimg/kw/{kw}/kwn/{kwn}/start/{next}/');}}</script>"
    ))
}

pub async fn words(State(state): State<AppState>, Query(params): Query<Params>) -> Response {
    qzone_words(state, params, 1070092, "word", "text", None, "words").await
}

pub async fn qianming(State(state): State<AppState>, Query(params): Query<Params>) -> Response {
    qzone_words(
        state,
        params,
        1019,
        "autograph",
        "word",
        Some(("keyword", json!("伤感"))),
        "qianming",
    )
    .await
}

pub async fn wangming(State(state): State<AppState>, Query(params): Query<Params>) -> Response {
    qzone_words(
        state,
        params,
        1025,
        "name",
        "word",
        Some(("keyword", json!("女生"))),
        "wangming",
    )
    .await
}

pub async fn net(State(state): State<AppState>) -> Response {
    // The old action depended on a PHP scraper class. Keep the update surface alive
    // using the same hot-comment schema via the public comments endpoint.
    netease(State(state)).await
}

pub async fn dog(State(state): State<AppState>) -> Response {
    let body = match fetch_json(&state, "https://v1.alapi.cn/api/dog?format=json").await {
        Some(v) if v.get("code").and_then(Value::as_i64) == Some(200) => {
            let text = v
                .pointer("/data/content")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_string();
            if text.is_empty() {
                "内部错误".to_string()
            } else {
                match insert_unique(&state.pool, "dog", "text", &text, &[])
                    .await
                    .as_str()
                {
                    "更新成功" => "写入成功".to_string(),
                    "数据重复" => "重复".to_string(),
                    _ => "失败".to_string(),
                }
            }
        }
        _ => "内部错误".to_string(),
    };
    text(format!(
        "{body}<script>window.onload=function(){{window.location.replace('https://api.gqink.cn/api/update/dog/');}}</script>"
    ))
}

async fn qzone_words(
    state: AppState,
    params: Params,
    feed_tag_id: i64,
    table: &'static str,
    column: &'static str,
    extra: Option<(&'static str, Value)>,
    action: &'static str,
) -> Response {
    let start = params
        .get("start")
        .and_then(|v| v.parse::<i64>().ok())
        .unwrap_or(0);
    let next = start + 100;
    let url = format!(
        "http://api4.qzone.cc/feed/default/get-new-feed-by-feed-tag?feed_tag_id={feed_tag_id}&start={start}&len=100"
    );
    let inserted = match fetch_json(&state, &url).await {
        Some(v) => {
            let mut inserted = 0;
            if let Some(items) = v.get("data").and_then(Value::as_array) {
                for item in items {
                    let Some(line) = value_string(item.get("text")) else {
                        continue;
                    };
                    let extras = extra
                        .as_ref()
                        .map(|(k, v)| vec![(*k, v.clone())])
                        .unwrap_or_default();
                    let extras = extras
                        .iter()
                        .map(|(k, v)| (*k, v.clone()))
                        .collect::<Vec<_>>();
                    if insert_unique_owned(&state.pool, table, column, &line, &extras).await
                        == "更新成功"
                    {
                        inserted += 1;
                    }
                }
                format!("写入{inserted}条成功")
            } else {
                "内部错误".to_string()
            }
        }
        None => "内部错误".to_string(),
    };
    text(format!(
        "<title>{inserted}</title>{inserted}<script>window.onload=function(){{window.location.replace('https://api.gqink.cn/api/update/{action}/start/{next}/');}}</script>"
    ))
}

async fn insert_hot_comment(pool: &MySqlPool, data: &Value) -> String {
    let comment = value_string(data.get("content")).unwrap_or_default();
    if comment.is_empty() {
        return "内部错误".into();
    }
    let exists = sqlx::query("SELECT id FROM `hot` WHERE comment_content=? LIMIT 1")
        .bind(&comment)
        .fetch_optional(pool)
        .await
        .ok()
        .flatten()
        .is_some();
    if exists {
        return "数据重复".into();
    }
    let inserted = sqlx::query(
        "INSERT INTO `hot` (name, images, author, mp3_url, comment_nickname, comment_content, love) VALUES (?, ?, ?, ?, ?, ?, ?)",
    )
    .bind(value_string(data.get("name")).unwrap_or_default())
    .bind(value_string(data.get("picurl")).unwrap_or_default())
    .bind(value_string(data.get("artistsname")).unwrap_or_default())
    .bind(value_string(data.get("url")).unwrap_or_default())
    .bind(value_string(data.get("nickname")).unwrap_or_default())
    .bind(comment)
    .bind(520_i64)
    .execute(pool)
    .await
    .is_ok();
    if inserted {
        "更新成功".into()
    } else {
        "更新失败".into()
    }
}

async fn insert_unique(
    pool: &MySqlPool,
    table: &str,
    unique_column: &str,
    unique_value: &str,
    extra: &[(&str, Value)],
) -> String {
    let owned = extra
        .iter()
        .map(|(k, v)| (*k, v.clone()))
        .collect::<Vec<_>>();
    insert_unique_owned(pool, table, unique_column, unique_value, &owned).await
}

async fn insert_unique_owned(
    pool: &MySqlPool,
    table: &str,
    unique_column: &str,
    unique_value: &str,
    extra: &[(&str, Value)],
) -> String {
    let Some(table_sql) = quote_table(table) else {
        return "内部错误".into();
    };
    let column_sql = quote_ident(unique_column);
    let exists_sql = format!("SELECT id FROM {table_sql} WHERE {column_sql}=? LIMIT 1");
    let exists = sqlx::query(&exists_sql)
        .bind(unique_value)
        .fetch_optional(pool)
        .await
        .ok()
        .flatten()
        .is_some();
    if exists {
        return "数据重复".into();
    }

    let mut columns = vec![column_sql];
    let mut placeholders = vec!["?".to_string()];
    for (column, _) in extra {
        columns.push(quote_ident(column));
        placeholders.push("?".to_string());
    }
    let insert_sql = format!(
        "INSERT INTO {table_sql} ({}) VALUES ({})",
        columns.join(","),
        placeholders.join(",")
    );
    let mut query = sqlx::query(&insert_sql).bind(unique_value.to_string());
    for (_, value) in extra {
        query = bind_json_value(query, value);
    }
    if query.execute(pool).await.is_ok() {
        "更新成功".into()
    } else {
        "更新失败".into()
    }
}

fn bind_json_value<'q>(
    query: sqlx::query::Query<'q, sqlx::MySql, sqlx::mysql::MySqlArguments>,
    value: &Value,
) -> sqlx::query::Query<'q, sqlx::MySql, sqlx::mysql::MySqlArguments> {
    if let Some(v) = value.as_i64() {
        query.bind(v)
    } else if let Some(v) = value.as_u64() {
        query.bind(v)
    } else if let Some(v) = value.as_f64() {
        query.bind(v)
    } else if let Some(v) = value.as_bool() {
        query.bind(v)
    } else {
        query.bind(value.as_str().unwrap_or_default().to_string())
    }
}

async fn fetch_json(state: &AppState, url: &str) -> Option<Value> {
    let text = state.http.get(url).send().await.ok()?.text().await.ok()?;
    serde_json::from_str(text.trim_start_matches('\u{feff}')).ok()
}

fn value_string(value: Option<&Value>) -> Option<String> {
    match value? {
        Value::String(s) => Some(s.clone()),
        Value::Number(n) => Some(n.to_string()),
        Value::Bool(v) => Some(if *v { "1" } else { "0" }.to_string()),
        _ => None,
    }
}

fn quote_table(table: &str) -> Option<String> {
    let table = table.trim();
    if table.is_empty()
        || table
            .chars()
            .any(|c| c == '`' || c == ';' || c.is_control())
    {
        None
    } else {
        Some(format!("`{table}`"))
    }
}

fn quote_ident(column: &str) -> String {
    format!("`{}`", column.replace('`', ""))
}

fn text(body: impl Into<String>) -> Response {
    (
        StatusCode::OK,
        [(header::CONTENT_TYPE, "text/html; charset=utf-8")],
        body.into(),
    )
        .into_response()
}
